typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,234 @@
+import { SamlangExpression } from '../ast/lang/samlang-expressions';
+import { IdentifierType } from '../ast/common/types';
+import { InterpretationContext, EMPTY } from './interpretation-context';
+import { Value, ObjectValue, IntValue, BoolValue, StringValue, FunctionValue, VariantValue, TupleValue } from './value';
+import { PanicException } from './panic-exception';
+
+export default class ExpressionInterpreter {
+  private printedCollector: string = "";
+
+  private blameTypeChecker = (message: string = ""): never => {
+    throw Error(message);
+  };
+
+  readonly eval = (expression: SamlangExpression, context: InterpretationContext = EMPTY): Value => {
+    switch (expression.__type__) {
+      case 'LiteralExpression': {
+        switch (expression.literal.type) {
+          case 'IntLiteral':
+            return { type: 'int', value: Number(expression.literal.value) };
+          case 'StringLiteral':
+            return { type: 'string', value: expression.literal.value };
+          case 'BoolLiteral':
+            return { type: 'bool', value: expression.literal.value };
+        };
+      }
+      case 'ThisExpression':
+        return context.localValues.get("this") ?? this.blameTypeChecker("Missing `this`");
+      case 'VariableExpression':
+        return context.localValues.get(expression.name) ?? this.blameTypeChecker(`Missing variable ${expression}`);
+      case 'ClassMemberExpression':
+        return context.classes.get(expression.className)?.functions?.get(expression.memberName) ?? this.blameTypeChecker();
+      case 'TupleConstructorExpression':
+        return { type: 'tuple', tupleContent: expression.expressions.map(e => this.eval(e, context)) }
+      case 'ObjectConstructorExpression':
+        const objectContent = new Map<string, Value>()
+        expression.fieldDeclarations.forEach((declaration) => {
+          if (declaration.expression) {
+            objectContent.set(declaration.name, this.eval(declaration.expression, context));
+          }
+        });
+        return { type: 'object', objectContent: objectContent };
+      case 'VariantConstructorExpression':
+        return { type: 'variant', tag: expression.tag, data: this.eval(expression.data, context) };
+      case 'FieldAccessExpression': {
+        const thisValue = this.eval(expression.expression, context) as ObjectValue;
+        return thisValue.objectContent.get(expression.fieldName) ?? this.blameTypeChecker();
+      }
+      case 'MethodAccessExpression': {
+        const { identifier } = expression.expression.type as IdentifierType;
+        const thisValue = this.eval(expression.expression, context);
+        const methodValue = context.classes.get(identifier)?.methods?.get(expression.methodName) ?? this.blameTypeChecker();
+        const newCtx = { ...context };
+        newCtx.localValues = context.localValues.set("this", thisValue);
+        // should context not be readonly?
+        methodValue.context = newCtx;
```

Comment:
`FunctionValue`'s context should not be readonly, since you need to patch it like this.